petastorm/reader_impl/worker_loop.py

```diff
@@ -138,27 +180,38 @@ def worker_loop(epochs_generator, loading_pool, loader, decoding_pool, decoder,
             while True:
                 try:
                     next_decoded = next(decoded)
-                    stats['rows_read_from_decoder'] += 1
+                    stats['row_batches_read_from_decoder'] += 1
                     already_decoded.add(next_decoded)
                 except (concurrent.futures.TimeoutError, StopIteration):
                     break
 
-            # follow the order of futures in the `in_decoding` list
+            stats['in_decoding_size'] = len(in_decoding)
+
             updated_in_decoding = []
             for i, was_decoding in enumerate(in_decoding):
                 if was_decoding in already_decoded:
                     stats['rows_written_to_output_queue'] += 1
                     while not stop_event.is_set():
                         try:
-                            output_queue.put(was_decoding.result(), block=False)
+                            result_buffer = was_decoding.result()
+                            if row_serializer:
+                                was_decoding_result = row_serializer.deserialize(result_buffer)
+                            else:
+                                was_decoding_result = result_buffer
+
+                            output_queue.put(was_decoding_result, block=False)
+                            stats['rows_written_to_output_queue'] += len(was_decoding_result)
+                            from_decoding.update(stats['rows_written_to_output_queue'])
                             break
                         except queue.Full:
-                            sleep(_LOOP_IDLE_TIME_SEC)
+                            raise RuntimeError('Do not set Queue\'s maxsize. The size of the queue is managed by '
```

Comment:
I cant immediately see what is managing the size of the output queue. Is there something stopping it from growing indefinitely?

Reply:
Fixed. All the control mechanism of the worker_loop becomes tricky and probably is fragile. Will need to think how to improve it.